### PR TITLE
sparse gradcheck: reparametrize some tests to remove masked=True

### PR DESCRIPTION
**Pull request by nikitaved**
Most of the sparse functions that work with sparse tensors assume that sparse is an optimization, so a green check with `masked=False` will imply success with `masked=True`. Functions that assume the sparse semantics and do not explicitly ignore grads outside of the sparse pattern can be re-parametrized with `torch.sparse_mask` so that the gradcheck succeeds when `masked=False`. Hence, we can remove `masked=True` altogether. 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99298
* #99043
* __->__ #98490
* #104341

cc @alexsamardzic @pearu @cpuhrsch @amjames @bhosmer @ezyang @albanD @zou3519 @gqchen @soulitzer @Lezcano @Varal7

### CODE REVIEW

File: test/test_sparse.py [hunk 1]
```diff
@@ -1798,24 +1798,27 @@ def test_shape(sparse_dims, nnz, with_size):
...
                 D_sum = D.sum(td)
                 self.assertEqual(S_sum.to_dense() if S_sum.is_sparse else S_sum, D_sum)
 
                 def fn(S):
                     res = torch.sparse.sum(S, td)
-                    return res.to_dense(masked_grad=True)
-                gradcheck(fn, (S,), masked=True)
+                    if res.layout != torch.strided:
+                        res = res.sparse_mask(res)
+                    return res.to_dense(masked_grad=False)
```
Comment (nikitaved) [inline test/test_sparse.py:1820]: Once `sparse_mask` supports CSR/CSC, it would make sense to move it inside the `torch.sparse.sum`'s body.

**Comment by pearu** [comment]
Hmm, removing `masked=True` will basically remove tests for `gradcheck(..., masked=True)` functionality. Since the functionality has not been deprecated, this PR might be premature.

**Comment by pearu** [comment]
> Most of the sparse functions that work with sparse tensors assume that sparse is an optimization, so a green check with masked=False will imply success with masked=True.

Consider the function `torch.mm` that supports sparse inputs and it should use non-masked semantics. However, your statement does not hold for the following example case:
```python
>>> a = torch.tensor([[0, 1], [2, 3]], dtype=torch.float64).to_sparse().requires_grad_(True)
>>> torch.autograd.gradcheck(lambda x: torch.mm(x, x).to_dense(masked_grad=False), (a,), masked=False)
True
>>> torch.autograd.gradcheck(lambda x: torch.mm(x, x).to_dense(masked_grad=True), (a,), masked=True)
<snip>
torch.autograd.gradcheck.GradcheckError: Jacobian mismatch for output 0 with respect to input 0,
numerical:tensor([[0.0000, 0.0000, 0.0000, 0.0000],
        [2.0000, 3.0000, 0.0000, 2.0000],
        [1.0000, 0.0000, 3.0000, 1.0000],
        [0.0000, 1.0000, 2.0000, 6.0000]], dtype=torch.float64)
analytical:tensor([[0., 1., 2., 0.],
        [2., 3., 0., 2.],
        [1., 0., 3., 1.],
        [0., 1., 2., 6.]], dtype=torch.float64)
```
The statement holds when the input sparse tensor is a full tensor:
```
>>> a = torch.tensor([[10, 1], [2, 3]], dtype=torch.float64).to_sparse().requires_grad_(True)
>>> torch.autograd.gradcheck(lambda x: torch.mm(x, x).to_dense(masked_grad=True), (a,), masked=True)
True
```

Next, consider `torch.sparse.mm` that implements `mm` with masked semantics:
```python
>>> a = torch.tensor([[0, 1], [2, 3]], dtype=torch.float64).to_sparse().requires_grad_(True)
>>> torch.autograd.gradcheck(lambda x: torch.sparse.mm(x, x).to_dense(masked_grad=True), (a,), masked=True)
True
```
but with `masked=False`, the gradcheck will fail:
```python
>>> torch.autograd.gradcheck(lambda x: torch.sparse.mm(x, x).to_dense(masked_grad=False), (a,), masked=False)
<snip>
torch.autograd.gradcheck.GradcheckError: Jacobian mismatch for output 0 with respect to input 0,
numerical:tensor([[0.0000, 1.0000, 2.0000, 0.0000],
        [2.0000, 3.0000, 0.0000, 2.0000],
        [1.0000, 0.0000, 3.0000, 1.0000],
        [0.0000, 1.0000, 2.0000, 6.0000]], dtype=torch.float64)
analytical:tensor([[0., 0., 0., 0.],
        [2., 3., 0., 2.],
        [1., 0., 3., 1.],
        [0., 1., 2., 6.]], dtype=torch.float64)
```
unless the input sparse tensor is full:
```python
>>> a = torch.tensor([[10, 1], [2, 3]], dtype=torch.float64).to_sparse().requires_grad_(True)
>>> torch.autograd.gradcheck(lambda x: torch.sparse.mm(x, x).to_dense(masked_grad=False), (a,), masked=False)
True
```

Based on the above, I cannot confirm that `masked=True` can be removed.

**Comment by nikitaved** [comment]
@pearu, it can be one the gradients are properly mapped to the manifold with sparse_mask.

**Comment by pearu** [comment]
> it can be once the gradients are properly mapped to the manifold with sparse_mask. 
 sparse.mm is just a composition of torch.mm and sparse_mask

I agree. This is what masked tensor support should handle, that is, `torch.mm` on masked tensors is equivalent to `torch.mm` and `torch.sparse_mask` combination on value-mask pairs as you exemplified.

`torch.sparse.mm` on sparse tensors is equivalent to `torch.mm` on masked tensors where the masks are defined by the sparsity patterns of inputs. Recall, defining the mask via sparsity pattern is something we want to get rid of eventually.

Our aim is to deprecate/eliminate `torch.sparse.mm` (together with the corresponding backward function) in favor of supporting masked tensors in `torch.mm`. Once we have this support and have applied the necessary deprecations procedures to handle BC-breaking changes (e.g. defining `torch.sparse.mm` in terms of `torch.mm` and converting sparse tensors to masked tensors), the `masked` kw argument can safely be removed from gradcheck.

Until then, removing the usage of `masked` kw argument as in this PR, is premature, IMHO. Especially when this is achieved by removing tests that exercise the `masked=True` case. When removing such tests, we risk that any future changes to gradcheck may introduce undetectable bugs for the `gradcheck(..., masked=True)` support which we are not ready to drop yet.

In general, we should deprecate a feature before removing the corresponding tests. In this PR, tests are removed but deprecating the feature is not immediately possible because masked tensors support is not ready yet.

**Comment by nikitaved** [comment]
I am in no rush to merge. It serves as a proof of concept: `gradcheck` does not need `masked=True` as it complicates things, is error-prune when it comes to implementing backward formulas, and it gives a false feeling of correct gradients once it succeeds.

**Pull request closed without merging**